tensorboard/db.py

```diff
@@ -20,160 +20,492 @@
+          rowid INTEGER PRIMARY KEY,
+          tensor BLOB NOT NULL
+        )
+      ''')
+
+  def create_plugins_table(self):
+    """Creates the Plugins table.
+
+    This table exists to assign arbitrary IDs to TBPlugin names. These
+    IDs are handed out in monotonically increasing order, but that's OK,
+    because the size of this table will be extremely small. These IDs
+    will not be the same across TensorBoard installs.
+
+    It is assumed that once an ID is mapped to a name, that the mapping
+    will never change.
+
+    Fields:
+      plugin_id: Arbitrary 16-bit integer.
+      name: (Uniquely indexed) Arbitrary string which is the same as the
+          TBPlugin.plugin_name field.
+    """
+    with self._cursor() as c:
+      c.execute('''\
+        CREATE TABLE IF NOT EXISTS Plugins (
+          plugin_id INTEGER PRIMARY KEY,
+          name TEXT NOT NULL
+        )
+      ''')
+
+  def create_plugins_table_name_index(self):
+    """Uniquely indexes the name field on the plugins table."""
+    with self._cursor() as c:
+      c.execute('''\
+        CREATE UNIQUE INDEX IF NOT EXISTS PluginsNameIndex
+        ON Plugins (name)
+      ''')
+
+  def create_event_logs_table(self):
+    """Creates the EventLogs table.
+
+    Event logs are files written to disk by TensorFlow via FileWriter,
+    which uses PyRecordWriter to output records containing binary-encoded
+    tf.Event protocol buffers.
+
+    This table is used by FileLoader to track the progress of files being
+    loaded off disk into the database.
+
+    Each time FileLoader runs a transaction committing events to the
+    database, it updates the offset field.
+
+    Fields:
+      rowid: An arbitrary event_log_id in the first 29 bits, and the
+          run_id in the higher 29 bits.
+      run_id: A reference to the id field of the associated row in the
+          runs table. Must be the same as what's in those rowid bits.
+      path: The basename of the path of the event log file. It SHOULD be
+          formatted: events.out.tfevents.UNIX_TIMESTAMP.HOSTNAME[SUFFIX].
+      offset: The byte offset in the event log file *after* the last
+          successfully committed event record.
+    """
+    with self._cursor() as c:
+      c.execute('''\
+        CREATE TABLE IF NOT EXISTS EventLogs (
+          rowid INTEGER PRIMARY KEY,
+          run_id INTEGER NOT NULL,
+          path VARCHAR(255) NOT NULL,
+          offset INTEGER NOT NULL
+        )
+      ''')
+
+  def create_event_logs_table_path_index(self):
+    """Uniquely indexes the (name, path) fields on the event_logs table."""
+    with self._cursor() as c:
+      c.execute('''\
+        CREATE UNIQUE INDEX IF NOT EXISTS EventLogsPathIndex
+        ON EventLogs (run_id, path)
+      ''')
+
+  def _cursor(self):
+    return contextlib.closing(self._db_conn.cursor())  # type: sqlite3.Cursor
+
+
+class Id(object):
+  """Utility class for unsigned fixed bit IDs."""
+
+  def __init__(self, name, bits):
```

Comment:
I'm surprised that you don't want to dependency-inject a parameter `random_=None` with `self._random = random_ or random.Random()`?